rfcs/FINEGRAINED_NOTIFICATIONS.md

```diff
@@ -0,0 +1,122 @@
+# Finegrained notifications
+
+## Summary
+
+This document describes an API proposal for a finegrained collection for Realm Java. Object-level notifications will be covered in another document.
+
+## Motivation
+
+Realm wants to promote a reactive pattern to working with your data. It should be possible to define a query for any change you are interested in and just depend on Realm figuring out when that change has happend and notify you. 
+
+This is interesting for a number of use cases:
+
+1) I want to animate or refresh a RecyclerView when elements are added, moved, modified or deleted in a `RealmResults`.
+2) I want to be notified about the result of an query changing so I can refresh my UI.
+3) I want my changelisteners to _only_ trigger when there is a relevant change.
+
+While you in some use cases probably want to compare old and new data, we have not yet found a compelling enough use case to support this for collections.
+Single object notifications will have this functionality.
+
+## Current situation
+
+- Right now Realm Java changelisteners are "table-based", which means that a change to any element in a table will trigger all changelisteners registered on that table, even though the element wasn't included in the query. This results in a lot of un-needed change events being triggered.
+
+- Single object change listeners work the same way, which is even worse.
+
+- Changelisteners are not supported on RealmList
+
+- Cocoa has support for fine-grained collection changes.
+
+- `Realm` / `RealmResults` / `RealmObject` all have methods like `addChangeListener/removeChangeListener/removeAllListeners`.
+
+
+## Detailed design
+
+Add these new interfaces and class:
+
+```
+// New generic interface for generic changes (= our current changelisteners)
+// All relevant classes should implement this: Realm, RealmResults, RealmList, RealmObject, DynamicRealm, DynamicRealmObject
+// TODO: Is Observable the best name?
```

Comment:
I like the name but feel it may confuse user with RxJava's same name interface. But I don't have a better suggestion.